### DISCUSSION TITLE
Multistep forecast models?

### DISCUSSION THREAD

**Discussion by caseytomlin** [Nixtla/mlforecast#68] [Q&A] [answered]
Is it the case that all models provide strictly recursive/autoregressed forecasts? If so, is it on the roadmap to support "direct" multistep models, like as wrappers around any sklearn models that support vector output? If this is already supported, could someone point me to a usage example? Thanks!

**Comment by jmoralez** [comment, marked as answer]
Hi @caseytomlin, can you provide an example of what you mean? Do these models take the same input and produce the predictions for the whole window? If that is the case you could use mlforecast to do the feature engineering with [MLForecast.preprocess](https://nixtla.github.io/mlforecast/forecast.html#mlforecast.preprocess) and then fit those models to the output.

**Reply by caseytomlin** [reply]
Maybe my question is more about the preprocessing functionality actually. For example, `xgb.XGBRegressor().fit(X, y)` supports higher-dimensional `y`. I could imagine passing an integer argument like `horizon_length` to the preprocessor that would be used to generate vector targets `y` of length `horizon_length` (plus restrict the allowed lags/other rolling features).

**Reply by jmoralez** [reply]
I believe those fit a model to each horizon, right? So it'd be using a model for each horizon instead of the recursive strategy? We do have that on the roadmap.

**Reply by jmoralez** [reply]
We've added that approach in #80. Please let us know if that fits your use case.

**Reply by caseytomlin** [reply]
Thanks, @jmoralez ! This is a great addition, though I actually had something different in mind. There are three general methods I can think of:

1. Recursive: train one 1-step model, predict `max_horizon = n` by recursively appending predictions as input (already implemented)
2. Multi-model: implemented in [#80](https://github.com/Nixtla/mlforecast/pull/80)
3. Single model, multi-output: trained on examples `(X, y)` where each `y` has length `max_horizon`

So only 3. is missing as far as I understand.

**Reply by jmoralez** [reply]
Do you have an example model that does that? xgboost does 2. internally

**Reply by caseytomlin** [reply]
ah ok, this was a misunderstanding on my part - apologies for the confusion and thanks for the clarification. It seems [what I had in mind](https://github.com/dmlc/xgboost/pull/8616) is still in active development.

**Comment by DAT-FYAYC** [comment]
To me the documentation of recursive strategy is unclear. So is the documentation for Multi-output model.
As stated in the documentation/code: "By default mlforecast uses the recursive strategy, i.e. a model is trained to predict the next value and if we're predicting several values we do it one at a time and then use the model's predictions as the new target, recompute the features and predict the next step."
How exactly is recomputing of features done? As far as my use case shows the features are simply duplicated from the previous time step.

How are "dynamic_dfs" in the "predict" method handelt in general as well as related to recursive and multi-model?

**Reply by DAT-FYAYC** [reply]
@jmoralez as far as I can see the "one model for each horizon" strategy is not yet implemented for the method "cross_validation". Is there any plan to do so?

**Reply by jmoralez** [reply]
With the recursive approach if you're using lag1 as a feature then for the next timestep you can use the last value of your target, however for the next one you don't have the actual value so you use your previous prediction as the lag1 and so on. This is only done for features that depend on the target, if you're only using static features like product id or something they will be repeated.

The dynamic_dfs are used to pass in features that don't depend on the target but change over time, e.g. product prices, holidays and so on. For those types of features both approaches produce the same values because they're just taken from the argument and joined with the inputs in each timestep.

I missed the CV when adding that, I've created #84 to include it.

**Reply by DAT-FYAYC** [reply]
@jmoralez thanks for the quick answer and the quick fix. I just a had a look at "LightGBMCV" and here it seems that changes also need to be made.

Concerning the use of features with MLForecast. In training all features will be used by the model that are passed within the fit method by the passed dataframe:
- For prediction with the **recursive strategy** with horizon>=2: lagged based features will be calculated and passed recursively based on **last point in time (recursively)** and static features are static anyways. For dynamic features the dataframe with the corresponding feature of **size (len_horizon * dynamic_features)** must be passed to provide the proper information. Do I understand it correctly?
- For prediction with the **multi-model strategy** with horizon>=2: lagged based features will be calculated and passed based **only on last point in time** and static features are static anyways. For dynamic features the dataframe with the corresponding feature of **size (1 * dynamic_features)** must be passed to provide the proper information. Do I understand it correctly?

**Reply by jmoralez** [reply]
LightGBMCV is just meant as a tool to estimate performance and get hyperparameters, not predicting to the future. Also it currently supports only the recursive strategy within the folds.
For predicting you need to instantiate an MLForecast object, for which we provide `MLForecast.from_cv`

**Reply by DAT-FYAYC** [reply]
@jmoralez Thank you for the explanation.

Regarding LightGBMCV this makes sense to me.

Concerning predictions this is still unclear to me. If I have an instance of MLForecast object (regardless if it was trained via the fit method or via cv) and I want to do a prediction for new (unseen) data. I need to provide the input properly. Depending on the strategy the input should vary when using dynamic features. Since with the recursive strategy e.g. for a forecasting horizon=2, the dynamic features of 2 time steps must be provided. Which is not the case for the multi-model strategy. As described above. I tried to check the code but find it hard to verify my understanding.

Would appreciate it a lot if you could confirm or correct my understanding.

**Reply by jmoralez** [reply]
That's correct. Since the recursive strategy computes predictions one at a time you need to provide the dynamic features for the next `horizon` timesteps, whereas for the multi-model you only need to provide the values for the next timestep, because that approach computes all the horizon in one go. If you provide more they will be ignored though, so you should be able to use both approaches passing the dynamic dfs for all the horizon.

**Reply by DAT-FYAYC** [reply]
Great, thanks a lot for the clarification.

**Reply by jmoralez** [reply]
We'd welcome a PR that clarifies this in the docs if you have time. Also please open an issue if you run into any problems.

**Reply by DAT-FYAYC** [reply]
@jmoralez I am happy to create a PR that clarifies this in the docs once I tested it.
